ctapipe/image/extractor.py

```diff
@@ -1285,3 +1290,373 @@ def __call__(
+    upsampling : int
+        Upsampling factor to use (>= 1); if > 1, the input waveforms are resampled
+        at upsampling times their original sampling rate.
+    pole_zero : float
+        Deconvolution parameter obtained from `deconvolution_parameters(...)`.
+
+    Returns
+    -------
+    deconvolved_waveforms : ndarray
+        Deconvolved and upsampled waveforms stored in a numpy array.
+        Shape: (n_pix, upsampling * n_samples)
+    """
+    deconvolved_waveforms = np.atleast_2d(waveforms) - np.atleast_2d(baselines).T
+    deconvolved_waveforms[:, 1:] -= pole_zero * deconvolved_waveforms[:, :-1]
+    deconvolved_waveforms[:, 0] = 0
+    if upsampling > 1:
+        return filtfilt(
+            np.ones(upsampling),
+            upsampling,
+            np.repeat(deconvolved_waveforms, upsampling, axis=-1),
+        )
+    return deconvolved_waveforms
+
+
+@guvectorize(
+    [
+        (float64[:], int64, float64, float32[:]),
+        (float32[:], int64, float64, float32[:]),
+    ],
+    "(s),(),()->()",
+    nopython=True,
+    cache=True,
+)
+def adaptive_centroid(waveforms, peak_index, rel_descend_limit, cog):
+    """
+    Calculates the pulse centroid for all samples down to rel_descend_limit * peak_amplitude.
+
+    The ret argument is required by numpy to create the numpy array which is
+    returned. It can be ignored when calling this function.
+
+    Parameters
+    ----------
+    waveforms : ndarray
+        Waveforms stored in a numpy array.
+        Shape: (n_pix, n_samples)
+    peak_index : ndarray or int
+        Peak index for each pixel.
+    rel_descend_limit : ndarray or float
+        Fraction of the peak value down to which samples are accumulated in the centroid calculation.
+    cog : ndarray
+        Return argument for ufunc (ignore)
+        Returns the peak centroid in units "samples"
+
+    Returns
+    -------
+    cog : ndarray
+        Peak centroid in units "samples"
+    """
+    cog[0] = peak_index  # preload in case of errors
+
+    nsamples = waveforms.size
+    if nsamples == 0:
+        return
+
+    peak_index = max(0, min(peak_index, nsamples - 1))
+    peak_amplitude = waveforms[peak_index]
+    if peak_amplitude < 0.0:
+        return
+
+    descend_limit = rel_descend_limit * peak_amplitude
+
+    sum = float64(0.0)
+    jsum = float64(0.0)
+    j = peak_index
+    while j >= 0 and waveforms[j] > descend_limit:
+        sum += waveforms[j]
+        jsum += j * waveforms[j]
+        j -= 1
+        if waveforms[j] > peak_amplitude:
+            descend_limit = rel_descend_limit * peak_amplitude
+
+    j = peak_index + 1
+    while j < nsamples and waveforms[j] > descend_limit:
+        sum += waveforms[j]
+        jsum += j * waveforms[j]
+        j += 1
+        if waveforms[j] > peak_amplitude:
+            descend_limit = rel_descend_limit * peak_amplitude
+
+    if sum != 0.0:
+        cog[0] = jsum / sum
+
+
+class FlashCamExtractor(ImageExtractor):
```

Comment:
Missing docstring. Looking at the complexity, this needs a lot of information and references. At least one to the design of the flashcam electronics that make this necessary and a general description of the approach.